src/validators/mod.rs

```diff
@@ -317,28 +319,10 @@ fn build_specific_validator<'a, T: BuildValidator>(
         // * if the ref is used inside itself, we have to store the validator in slots,
         //   and return a DefinitionRefValidator - two step process with `prepare_slot` and `complete_slot`
         // * if the ref is used elsewhere, we want to clone it each time it's used
-        if build_context.ref_used(&schema_ref) {
-            // the ref is used somewhere
-            // check the ref is unique
-            if build_context.ref_already_used(&schema_ref) {
-                return py_err!("Duplicate ref: `{}`", schema_ref);
-            }
-
-            return if build_context.ref_used_within(schema_dict, &schema_ref)? {
-                // the ref is used within itself, so we have to store the validator in slots
-                // and return a DefinitionRefValidator
-                let slot_id = build_context.prepare_slot(schema_ref)?;
-                let inner_val = T::build(schema_dict, config, build_context)?;
-                let name = inner_val.get_name().to_string();
-                build_context.complete_slot(slot_id, inner_val)?;
-                Ok(definitions::DefinitionRefValidator::from_id(slot_id, name))
-            } else {
-                // ref is used, but only out side itself, we want to clone it everywhere it's used
-                let validator = T::build(schema_dict, config, build_context)?;
-                build_context.store_reusable(schema_ref, validator.clone());
-                Ok(validator)
-            };
-        }
+        let validator_id = build_context.get_ref_id(schema_ref);
+        let inner_val = T::build(schema_dict, config, build_context)?;
+        build_context.add_definition(validator_id, inner_val)?;
+        return Ok(DefinitionRefValidator::from_id(validator_id).into());
```

Comment:
The big question here is: what is the cost of this change? In theory, it's large. As you can see anything with a `ref` permanently becomes a `DefinitionRefValidator`, which also always use slots so they'll be "slow".

However this is not what needs to happen in practice: it would be _quite easy_ to, in Python, remove any unused `refs`. Certainly _much much_ easier than maintaining the complex system of refs and slots and whatnot that we had here. And in the end, it will be just as efficient because we can give `pydantic-core` the most efficient arrangement of references to begin with instead of trying to jump through hoops in Rust to get the borrow checker to accept circular references.
If we do that in Pydantic we can even _enforce_ it here if we want to.

Reply:
seems fine to me, but are we sure it's okay for core-schema's validator? If that's not effected, I'm fine to move it to pydantic.

The other option is to have another pass at reference counting in pydantic-core after this is merged?